rpctest/rpc_harness.go

```diff
@@ -0,0 +1,395 @@
+
+	if handlers == nil {
+		handlers = &btcrpcclient.NotificationHandlers{}
+	}
+
+	// If a handler for the OnBlockConnected/OnBlockDisconnected callback
+	// has already been set, then we create a wrapper callback which
+	// executes both the currently registered callback, and the mem
+	// wallet's callback.
+	if handlers.OnBlockConnected != nil {
+		obc := handlers.OnBlockConnected
+		handlers.OnBlockConnected = func(hash *chainhash.Hash, height int32, t time.Time) {
+			obc(hash, height, t)
+			wallet.ingestBlock(hash, height, t)
+		}
+	} else {
+		// Otherwise, we can claim the callback ourselves.
+		handlers.OnBlockConnected = wallet.ingestBlock
+	}
+	if handlers.OnBlockDisconnected != nil {
+		obd := handlers.OnBlockConnected
+		handlers.OnBlockDisconnected = func(hash *chainhash.Hash, height int32, t time.Time) {
+			obd(hash, height, t)
+			wallet.unwindBlock(hash, height, t)
+		}
+	} else {
+		handlers.OnBlockDisconnected = wallet.unwindBlock
+	}
+
+	h := &Harness{
+		handlers:       handlers,
+		node:           node,
+		maxConnRetries: 20,
+		testNodeDir:    nodeTestData,
+		ActiveNet:      activeNet,
+		nodeNum:        nodeNum,
+		wallet:         wallet,
+	}
+
+	// Track this newly created test instance within the package level
+	// global map of all active test instances.
+	testInstances[h.testNodeDir] = h
+
+	return h, nil
+}
+
+// SetUp initializes the rpc test state. Initialization includes: starting up a
+// simnet node, creating a websockets client and connecting to the started
+// node, and finally: optionally generating and submitting a testchain with a
+// configurable number of mature coinbase outputs coinbase outputs.
+func (h *Harness) SetUp(createTestChain bool, numMatureOutputs uint32) error {
+	var err error
+
+	// Start the btcd node itself. This spawns a new process which will be
+	// managed
+	if err = h.node.start(); err != nil {
+		return err
+	}
+	if err := h.connectRPCClient(); err != nil {
+		return err
+	}
+
+	h.wallet.start()
+
+	// Ensure the wallet properly dispatches our registered call-back for
+	// each new block. Otherwise, the memWallet won't function properly.
+	if err := h.Node.NotifyBlocks(); err != nil {
+		return err
+	}
+
+	// Create a test chain with the desired number of mature coinbase
+	// outputs.
+	if createTestChain {
+		numToGenerate := (uint32(h.ActiveNet.CoinbaseMaturity) +
+			numMatureOutputs)
+		_, err := h.Node.Generate(numToGenerate)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Block until the wallet has fully synced up to the tip of the main
+	// chain.
+	_, height, err := h.Node.GetBestBlock()
+	if err != nil {
+		return err
+	}
+	ticker := time.NewTicker(time.Millisecond * 100)
+out:
+	for {
+		select {
+		case <-ticker.C:
+			walletHeight := h.wallet.syncedHeight()
+			if walletHeight == height {
+				break out
+			}
+		}
+	}
+
+	return nil
+}
+
+// TearDown stops the running rpc test instance. All created processes are
+// killed, and temporary directories removed.
```

Comment:
Should it be made clear that `SetUp` and `TearDown` have to be called from the same goroutine since they aren't concurrent safe?  That is almost always going to be the case anyways, but since this is intended to be usable by others as well, that's important information.